### PR TITLE
Wait for gossip to start after starting a new conductor

### PR DESCRIPTION
**Pull request by ThetaSinner**
### Summary

- Check that working directories are not being used another conductor (across tests) by checking that the proposed directory is empty.
- Ensure that gossip has actually started before waiting for database consistency. This helps separate the two problems that happen on CI. Either gossip doesn't start or the database integration does not appear to finish.
- Make `ROUND_TIMEOUT ` a tuning parameter. The default 1 minute means that if kitsune talks to a conductor which is being shutdown first then the round will finish in a reasonable amount of time and try talking to another node.
- Add another tuning parameter `bootstrap_check_delay_backoff_multiplier` to prevent peer discovery from slowing down over time. This helps significantly speed up tests when running multiple gossip tests in parallel locally.
- There was an issue with `add_conductor` starting new local services. I have added another function to the sweet conductor batch which lets you add a conductor from config instead, which will re-use the local services that the batch is already using.
- There also appears to have been a change that was wider than intended to the sweet conductor batch to use local services. I've provided an option to run conductors and nothing else which it appears was the default before. Tests that rely on knowing when gossip starts should use this instead.

### TODO:
- [ ] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

**Comment by ThetaSinner** [comment]
I think these improvements are worth getting in although I haven't completely resolved the CI issues. The intel+macos tests seem to be pretty much stable now. The darwin+macos tests are still misbehaving when multiple test runs are in progress.

At least with these changes I don't have trouble running these tests locally. They are stable when run individually at least. Running with `nextest` they are mostly stable (some tests fail maybe 1/20 runs). When starting several test runs in parallel I still see errors like we get on CI.

I'm on the fence about re-enabling problem tests before they're completely fixed. Happy to take input on that. I'm slightly leaning towards disabling tests that were disabled on develop after the changes have been reviewed. I'd like to leave them in place so that reviewers can see how the builds are looking.

File: crates/holochain/src/sweettest/sweet_conductor.rs [hunk 1]
```diff
@@ -186,6 +192,11 @@ impl SweetConductor {
         Self::from_config(standard_config()).await
     }
 
+    /// Get the rendezvous config that this conductor is using, if any
+    pub fn get_rendezvous_config(&self) -> Option<DynSweetRendezvous> {
+        self._rendezvous.clone()
```
Comment (neonphog) [inline crates/holochain/src/sweettest/sweet_conductor.rs:197]: If we're using this now, let's remove the underscore.

```suggestion
        self.rendezvous.clone()
```

**Review by neonphog — APPROVED** [review]
Woot, looks awesome. Nice work!

File: crates/holochain/src/sweettest/sweet_conductor.rs [hunk 2]
```diff
@@ -542,6 +553,53 @@ impl SweetConductor {
...
+    /// Wait for at least one gossip round to have completed for the given cell
+    ///
+    /// Note that this is really a crutch. If gossip starts fast enough then this is unnecessary
+    /// but that doesn't necessarily happen. Waiting for gossip to have started before, for example,
+    /// waiting for something else like consistency is useful to ensure that communication has
+    /// actually started.
+    pub async fn require_initial_gossip_activity_for_cell(
+        &self,
+        cell: &SweetCell,
+        timeout: Duration,
```
Comment (maackle) [inline crates/holochain/src/sweettest/sweet_conductor.rs:566]: It's kind of misleading to be checking gossip activity for a cell, since gossip applies to the entire conductor. I see why you did it though: because `network_info` now requires an agent key (via #2183). I think this is unfortunate and we shouldn't need to specify an agent key to get gossip info. I'd like us (the team) to consider refactoring this, but just making note of it for now.

File: crates/holochain/src/sweettest/sweet_conductor_batch.rs [hunk 3]
```diff
@@ -178,6 +218,17 @@ impl SweetConductorBatch {
...
+
+    /// Wait for the first conductor to have started gossipping. If all conductors in this batch were
+    /// started at the same time, this should be enough to ensure that all conductors have started gossipping.
+    /// If other conductors are added later, separately check that they have started gossipping as required.
+    pub async fn require_initial_gossip_activity_for_cell(
+        &self,
+        cell: &SweetCell,
+        timeout: Duration,
+    ) {
+        SweetConductor::require_initial_gossip_activity_for_cell(&self.0[0], cell, timeout).await;
```
Comment (maackle) [inline crates/holochain/src/sweettest/sweet_conductor_batch.rs:230]: If not for the change added in #2183 (which I questioned in a comment above), it would be easy to wait for all conductors in parallel, but as it stands it would require including a list of cells from each conductor, which is pretty annoying. It seems that it would be good to wait for all of them, otherwise we're just decreasing our chances of not running into this problem, but not fully eliminating the chance.

File: crates/holochain/src/sweettest/sweet_conductor_batch.rs [hunk 4]
```diff
@@ -45,11 +47,34 @@ impl SweetConductorBatch {
...
+    }
+
     /// Create the given number of new SweetConductors, each with its own new TestEnvironments
     pub async fn from_standard_config(num: usize) -> SweetConductorBatch {
         Self::from_configs(std::iter::repeat_with(SweetConductorConfig::standard).take(num)).await
     }
 
+    /// Create the given number of new SweetConductors, each with its own new TestEnvironments.
+    /// No network infrastructure will be created, these are just local conductors that can discover each other through `exchange_peer_info`
+    pub async fn from_standard_local_config(num: usize) -> SweetConductorBatch {
```
Comment (maackle) [inline crates/holochain/src/sweettest/sweet_conductor_batch.rs:72]: I wonder if we can consider making this the standard, and the rendezvous stuff nonstandard? Most tests were written with explicit agent injection, which I think don't require the rendezvous stuff?

**Review by maackle — APPROVED** [review]
Looks good, though I hope to change this to wait for gossip to start on all nodes if (1) this crutch winds up being useful for a long time, (2) the fact of waiting for only one conductor winds up being flaky and (3) we can get network info for the entire node without specifying an agent.

File: crates/holochain/src/sweettest/sweet_conductor_batch.rs (hunk 4 above)
Comment (ThetaSinner) [reply crates/holochain/src/sweettest/sweet_conductor_batch.rs]: Yes that's probably safer than just moving stuff over to local only when it fails

File: crates/holochain/src/sweettest/sweet_conductor.rs (hunk 2 above)
Comment (ThetaSinner) [reply crates/holochain/src/sweettest/sweet_conductor.rs:572]: Agreed, happy to update this if the `network_info` interface changes. Equally happy to work on making that change

**Comment by ThetaSinner** [comment]
Rebased onto the new changes in the `sharded_gossip` tests
